### PR TITLE
Allow potentially using THP (transparent huge pages) in sorbet_ruby

### PR DESCRIPTION
**Pull request by maruth-stripe**
### Motivation

The Ruby interpreter outright disables THP since it might have bad interactions with dirty pages (i.e., if a single bit in a 2MB page is dirtied now the entire things needs to be invalidated/flushed instead of just 4K) w.r.t GC etc. However, we would like to experiment with enabling it for memory-intensive workloads 

### Test plan

See included automated tests.

### CODE REVIEW

File: third_party/ruby/thp.patch [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
++++ eval.c
+@@ -72,13 +72,6 @@ ruby_setup(void)
+ 
+     ruby_init_stack((void *)&state);
+ 
+-    /*
+-     * Disable THP early before mallocs happen because we want this to
+-     * affect as many future pages as possible for CoW-friendliness
+-     */
+-#if defined(__linux__) && defined(PR_SET_THP_DISABLE)
```
Comment (fable-stripe) [inline third_party/ruby/thp.patch:13]: I suggest adding 

```
if (strncmp(getenv("STRIPE_RUBY_ENABLE_THP"), "true", 4) != 0) {
...
}
```

here

File: third_party/ruby_externals.bzl [hunk 2]
```diff
@@ -59,6 +59,7 @@ def register_ruby_dependencies():
         strip_prefix = strip_prefix,
         build_file = ruby_build,
         patches = [
+            "@com_stripe_ruby_typer//third_party/ruby:thp.patch",
```
Comment (amytsai-stripe) [inline third_party/ruby_externals.bzl:62]: Can you also include the patch in 3.1 below?

Comment (jez) [inline third_party/ruby_externals.bzl:62]: Is it important that you put the patch first? We have traditionally put all new patches at the bottom.

Comment (maruth-stripe) [reply third_party/ruby_externals.bzl:62]: ack I'll put it at the bottom

**Comment by maruth-stripe** [comment]
Verified this works by running
` bazel-bin/external/sorbet_ruby_2_7/ruby -e "x = []; i = 1; while true; x << 'abc' * i; end"`

And then on doing `cat /proc/<pid>/status` got
```
THP_enabled:	0
```

On running `STRIPE_RUBY_ENABLE_THP="true" bazel-bin/external/sorbet_ruby_2_7/ruby -e "x = []; i = 1; while true; x << 'abc' * i; end"`

And then on doing `cat /proc/<pid>/status` got
```
THP_enabled:	1
```

**Comment by maruth-stripe** [comment]
^also reprod above with `sorbet_ruby_3_1`

**Review by jez — APPROVED** [review]

**Pull request merged**